tools/crm_rule.c

```diff
@@ -0,0 +1,296 @@
+#include <crm/msg_xml.h>
+#include <crm/pengine/rules_internal.h>
+#include <crm/pengine/status.h>
+
+#include <sys/stat.h>
+
+#define CMD_ERR(fmt, args...) do {          \
+        crm_warn(fmt, ##args);              \
+        fprintf(stderr, fmt "\n", ##args);  \
+    } while(0)
+
+enum crm_rule_mode {
+    crm_rule_mode_none,
+    crm_rule_mode_check
+} rule_mode = crm_rule_mode_none;
+
+static int crm_rule_check(pe_working_set_t *data_set, const char *rule_id, crm_time_t *effective_date);
+
+static struct crm_option long_options[] = {
+    /* Top-level Options */
+    {"help",       no_argument,       NULL, '?', "\tThis text"},
+    {"version",    no_argument,       NULL, '$', "\tVersion information"  },
+    {"verbose",    no_argument,       NULL, 'V', "\tIncrease debug output"},
+
+    {"-spacer-",   required_argument, NULL, '-', "\nModes (mutually exclusive):" },
+    {"check",      no_argument,       NULL, 'c', "\tCheck whether a rule is in effect" },
+
+    {"-spacer-",   required_argument, NULL, '-', "\nAdditional options:" },
+    {"date",       required_argument, NULL, 'd', "Whether the rule is in effect on a given date" },
+    {"rule",       required_argument, NULL, 'r', "The ID of the rule to check" },
+
+    {"-spacer-",   no_argument,       NULL, '-', "\nData:"},
+    {"xml-text",   required_argument, NULL, 'X', "Use argument for XML (or stdin if '-')"},
+
+    {"-spacer-",   required_argument, NULL, '-', "\n\nThis tool is currently experimental.",
+     pcmk_option_paragraph},
+    {"-spacer-",   required_argument, NULL, '-', "The interface, behavior, and output may "
+                                                 "change with any version of pacemaker.", pcmk_option_paragraph},
+
+    {0, 0, 0, 0}
+};
+
+static int
+crm_rule_check(pe_working_set_t *data_set, const char *rule_id, crm_time_t *effective_date)
+{
+    xmlNode *cib_constraints = NULL;
+    xmlNode *match = NULL;
+    xmlXPathObjectPtr xpathObj = NULL;
+    pe_eval_date_result_t result;
+    char *xpath = NULL;
+    int rc = pcmk_ok;
+    int max = 0;
+
+    /* Rules are under the constraints node in the XML, so first find that. */
+    cib_constraints = get_object_root(XML_CIB_TAG_CONSTRAINTS, data_set->input);
+
+    /* Get all rules matching the given ID, which also have only a single date_expression
+     * child whose operation is not 'date_spec'.  This is fairly limited, but it's hard
+     * to check expressions more complicated than that.
+     */
+    xpath = crm_strdup_printf("//rule[@id='%s']/date_expression[@operation!='date_spec']", rule_id);
+    xpathObj = xpath_search(cib_constraints, xpath);
+
+    max = numXpathResults(xpathObj);
+    if (max == 0) {
+        CMD_ERR("No rule found with ID=%s containing a date_expression", rule_id);
+        rc = -ENXIO;
+        goto bail;
+    } else if (max > 1) {
+        CMD_ERR("More than one date_expression in %s is not supported", rule_id);
+        rc = -EOPNOTSUPP;
+        goto bail;
+    }
```

Comment:
The following XML blob does not end up triggering the case of more than one date_expression.  I haven't figured out why yet.
```
<rsc_location id="cli-prefer-dummy" rsc="dummy" role="Started">
  <rule id="cli-prefer-rule-dummy" score="INFINITY" boolean-op="and">
    <date_expression id="cli-prefer-lifetime-end-dummy" operation="lt" end="2019-02-25 12:21:52 -05:00"/>                                                                             
    <date_expression id="cli-prefer-lifetime-end-dummy2" operation="lt" end="2019-03-25 12:21:52 -05:00"/>                                                                          
  </rule>
</rsc_location>
```